sdk/clientcore/core/src/main/java/io/clientcore/core/implementation/AccessTokenCache.java

```diff
@@ -0,0 +1,168 @@
+import io.clientcore.core.credentials.TokenRequestContext;
+import io.clientcore.core.instrumentation.logging.ClientLogger;
+import io.clientcore.core.instrumentation.logging.LogLevel;
+
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+/**
+ * A token cache that supports caching a token and refreshing it.
+ */
+public final class AccessTokenCache {
+    // The delay after a refresh to attempt another token refresh
+    private static final Duration REFRESH_DELAY = Duration.ofSeconds(30);
+    private static final String REFRESH_DELAY_STRING = String.valueOf(REFRESH_DELAY.getSeconds());
+
+    // the offset before token expiry to attempt proactive token refresh
+    private static final Duration REFRESH_OFFSET = Duration.ofMinutes(5);
+    // AccessTokenCache is a commonly used class, use a static logger.
+    private static final ClientLogger LOGGER = new ClientLogger(AccessTokenCache.class);
+    private final AtomicReference<AccessTokenCacheInfo> cacheInfo;
+    private final TokenCredential tokenCredential;
+    // Stores the last authenticated token request context. The cached token is valid under this context.
+    private TokenRequestContext tokenRequestContext;
+    private final Supplier<AccessToken> tokenSupplierSync;
+    private final Predicate<AccessToken> shouldRefresh;
+
+    // Used for sync flow.
+    private final Lock lock;
+
+    /**
+     * Creates an instance of RefreshableTokenCredential with default scheme "Bearer".
+     *
+     * @param tokenCredential the token credential to be used to acquire the token.
+     */
+    public AccessTokenCache(TokenCredential tokenCredential) {
+        Objects.requireNonNull(tokenCredential, "The token credential cannot be null");
+        this.tokenCredential = tokenCredential;
+        this.cacheInfo = new AtomicReference<>(new AccessTokenCacheInfo(null, OffsetDateTime.now()));
+        this.shouldRefresh = accessToken -> OffsetDateTime.now()
+            .isAfter(accessToken.getRefreshAt() == null
+                ? accessToken.getExpiresAt().minus(REFRESH_OFFSET)
+                : accessToken.getRefreshAt());
+        this.tokenSupplierSync = () -> tokenCredential.getToken(this.tokenRequestContext);
+        this.lock = new ReentrantLock();
+    }
+
+    /**
+     * Synchronously get a token from either the cache or replenish the cache with a new token.
+     *
+     * @param tokenRequestContext The request context for token acquisition.
+     * @param checkToForceFetchToken The flag indicating whether to force fetch a new token or not.
+     * @return The Publisher that emits an AccessToken
+     */
+    public AccessToken getToken(TokenRequestContext tokenRequestContext, boolean checkToForceFetchToken) {
```

Comment:
Not a huge fan of `checkToForceFetchToken` - is it better to call it `forceTokenFetch` or `replenishToken` or similar?